src/System.Private.CoreLib/src/System/Reflection/CustomAttribute.cs

```diff
@@ -1294,6 +1294,11 @@ internal static object[] GetCustomAttributes(RuntimeType type, RuntimeType caTyp
                 type = type.BaseType as RuntimeType;
             }
 
+            if (result.Count == 0)
+            {
+                return useObjectArray ? Array.Empty<object>() : caType.GetEmptyArray();
+            }
+
             object[] typedResult = CreateAttributeArrayHelper(arrayType, result.Count);
             Array.Copy(result.ToArray(), 0, typedResult, 0, result.Count);
```

Comment:
Instead of doing `result.ToArray()` here that allocates a temporary array, it should be better to just go over the list and assign one element at a time. Another option would be to use a bulk copy but that's likely more expensive to setup here since there are a very few attributes in typical case.

Reply:
Agreed on both counts - simple loop copy added here, perf is much better.